File: Python/RiverSwim.py
from typing import cast

from matplotlib.pyplot import grid
import MDP
import RLCore
import numpy as np

class riverswim(MDP.mdp[int, str]):
    def __init__(self, size: int, stayProbability: float, reverseProbability: float, leftReward: float, rightReward: float, discount: float):
        self.size = size
        self.stayProbability = stayProbability
        self.reverseProbability = reverseProbability
        self.leftReward = leftReward
        self.rightReward = rightReward
        states = [x for x in range(size)]
        actions = ['left ', 'right']
        MDP.mdp.__init__(self, states, actions, discount)

    def reward(self, s1, a) -> float:
        if (s1 == 0):
            return self.leftReward
        elif (s1 == (self.size-1)):
            return self.rightReward
        return 0

    def baseTransition(self, s1: int, a: str) -> 'list[tuple[int, float]]':
        if (s1 == 0 and a == 'left '):
            return [(0, 1.0)]
        elif (s1 == 0 and a == 'right'):
            return [(0, self.stayProbability + self.reverseProbability), (1, 1.0 - self.stayProbability - self.reverseProbability)]
        elif (s1 == (self.size-1) and a == 'right'):
            return [(s1, self.stayProbability + self.reverseProbability), (s1-1, 1.0 - self.stayProbability - self.reverseProbability)]
        elif (a == 'left '):
            return [(s1-1, 1.0)]
        else:
            return [(s1-1, self.reverseProbability), (s1, self.stayProbability),
            (s1+1, 1.0 - self.reverseProbability - self.stayProbability)]

def idxToAction(action: int) -> str :
    return ['left ', 'right'][action]


def printV(V: 'list[float]') :
    print(' '.join(["+{:.2f}".format(val) for val in V]).replace('+-', '-'))


def stateQToString(Q: 'list[list[float]]', state: int) -> str :
    actionIdx: int = cast(int, np.argmax(Q[state]))
    return idxToAction(actionIdx)


def printActionsFromQ(Q: 'list[list[float]]'):
    print(' '.join([stateQToString(Q, x) for x in range(len(Q))]))

def printActionsFromPolicy(policy: 'list[str]'):
    print(' '.join(policy))

File: old (F#)/RiverSwim.fs
module RiverSwim

open System
open QLearning
open RLCore

type State = int
type Action = 
    | Left = 0
    | Right = 1
type Reward = float

let stateToString (s: State) : string =
    $"({s |> int})"

let actionToString (a: Action) : string =
    match a with
    | Action.Left -> "Left "
    | Action.Right -> "Right"
    | _ -> ""

/// Gets all the states for a world of the specified size
/// It is simply all combinations of x and y as well as the absorptionMarker
let getStates size  =
    [0 .. size-1]

let getActions: Action list = [Action.Left; Action.Right]

let applyActionSimple action stateFrom : State =
    match action with
    | Action.Left -> stateFrom - 1
    | Action.Right -> stateFrom + 1
    | _ -> raise (ArgumentException("Error"))

let baseTransitionFunction size (reverseProbability: float)  (stayProbability: float) stateFrom action : (State * float) list =
    if (stateFrom = 0 && action = Action.Left) then 
        [0, 1]
    else if (stateFrom = 0 && action = Action.Right) then 
        [0, (stayProbability + reverseProbability); 1, (1.0 - stayProbability - reverseProbability)]
    else if (stateFrom = (size-1) && action = Action.Right) then
        [stateFrom, (stayProbability + reverseProbability) ; applyActionSimple Action.Left stateFrom, (1.0 - stayProbability - reverseProbability)]
    else if (action = Action.Left) then
        [applyActionSimple action stateFrom, 1.0]
    else
        [applyActionSimple Action.Left stateFrom, reverseProbability ; stateFrom, stayProbability ;
        applyActionSimple Action.Right stateFrom, (1.0 - reverseProbability - stayProbability)]
        

let transitionfunction size (reverseProbability: float)  (stayProbability: float) : State->Action->State->float =
    let bt = baseTransitionFunction size reverseProbability stayProbability
    RLCore.convertTransitionFunction bt

let transitionfunctionStochastic size reverseProbability  stayProbability: State->Action->State =
    let bt = baseTransitionFunction size reverseProbability stayProbability
    QLearning.convertTransitionFunctionStochastic bt

let rewardFunction (size: int) (livingReward: Reward) (rightReward: Reward) (leftReward: Reward) stateFrom action : Reward =
    if (stateFrom = 0) then
        leftReward
    else if (stateFrom = (size-1)) then
        rightReward
    else
        livingReward
        

let stateIndex size state : int = int state

let actionIndex (action: Action) : int = int action

let printV size (Q: float[,]) : unit =
    [for x in 0 .. (size-1) do getStateValue (stateIndex size) x Q |>
                                                    sprintf "+%.2f" ] |>
                                                    String.concat " " |>
                                                    fun s -> s.Replace("+-", "-") |> printfn "%s"


let stateQToString size (Q: float[,]) state: string =
    let idx = stateIndex size state
    Q[idx, *] |>
            Seq.mapi (fun i v -> i, v) |> Seq.maxBy snd |> fst |>
            LanguagePrimitives.EnumOfValue<int, Action> |>
            actionToString


let printActions size (Q: float[,]) : unit =
    [for x in 0..(size-1) do (stateQToString size Q  x)] |>
        String.concat " " |> printfn "%s" 
